crosscluster/logical: TestRandomStream failed

crosscluster/logical.TestRandomStream [failed](https://mesolite.cluster.engflow.com/invocations/default/8d24cc06-f2ff-41c6-90cd-d7af43d26fe7?testReportRun=4&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xvZ2ljYWw6bG9naWNhbF90ZXN0) on master @ [fcf4659dae90ef3d87f8457678b1e6727e3d69b7](https://github.com/cockroachdb/cockroach/commits/fcf4659dae90ef3d87f8457678b1e6727e3d69b7):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait
	  | 	pkg/util/ctxgroup/ctxgroup.go:151
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).close
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:461
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).ConsumerClosed
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:437
	  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Close
	  | 	pkg/sql/execinfra/processorsbase.go:779
	  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup
	  | 	pkg/sql/flowinfra/flow.go:678
	  | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup
	  | 	pkg/sql/rowflow/row_based_flow.go:490
	  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3
	  | 	pkg/sql/distsql_running.go:1003
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
	  | 	pkg/util/log/logcrash/crash_reporting.go:434
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Next
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:397
	  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
	  | 	pkg/sql/execinfra/base.go:193
	  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
	  | 	pkg/sql/execinfra/processorsbase.go:766
	  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
	  | 	pkg/sql/flowinfra/flow.go:574
	  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
	  | 	pkg/sql/distsql_running.go:1068
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).ingest.func7
	  | 	pkg/crosscluster/logical/logical_replication_job.go:332
	  | [...repeated from below...]
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | runtime.panicmem
	  | 	GOROOT/src/runtime/panic.go:262
	  | runtime.sigpanic
	  | 	GOROOT/src/runtime/signal_unix.go:925
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.getColumnSchema
	  | 	pkg/crosscluster/logical/replication_statements.go:42
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.newEventDecoder.func1
	  | 	pkg/crosscluster/logical/event_decoder.go:82
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).DescsTxn.func1
	  | 	pkg/sql/internal.go:1955
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
	  | 	pkg/sql/internal.go:2052
	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
	  | 	pkg/kv/txn.go:1139
	  | github.com/cockroachdb/cockroach/pkg/kv.runTxn
	  | 	pkg/kv/db.go:1114
	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
	  | 	pkg/kv/db.go:1077
	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
	  | 	pkg/kv/db.go:1052
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
	  | 	pkg/sql/internal.go:2039
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).DescsTxn
	  | 	pkg/sql/internal.go:1953
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.newEventDecoder
	  | 	pkg/crosscluster/logical/event_decoder.go:75
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.newCrudSqlWriter
	  | 	pkg/crosscluster/logical/sql_crud_writer.go:51
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).setupBatchHandlers
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:729
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).flushBuffer
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:785
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).handleStreamBuffer
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:646
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).handleEvent
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:529
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).consumeEvents
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:502
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Start.func2.1
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:359
	  | github.com/cockroachdb/cockroach/pkg/util/pprofutil.Do
	  | 	pkg/util/pprofutil/labels.go:38
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Start.func2
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:358
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Start.Group.GoCtx.func10
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_amd64.s:1693
	Wraps: (3) runtime error: invalid memory address or nil pointer dereference
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) runtime.errorString
	panic: found 1 short-living non-stopped monitors in internal-planner.resume-job-1134799122344574977
	internal-planner.resume-job-1134799122344574977 0 B
	    flow-95f64f6d 0 B
	

```
Stack: 

```
goroutine 30267 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc00cfa8210, {0x8abdf40, 0xc00a917360}, 0x1)
	pkg/util/mon/bytes_usage.go:781 +0x653
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	pkg/util/mon/bytes_usage.go:722
github.com/cockroachdb/cockroach/pkg/sql.newInternalPlanner.func1()
	pkg/sql/planner.go:521 +0x5f
panic({0x720d940?, 0xc0119e7e60?})
	GOROOT/src/runtime/panic.go:783 +0x132
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc0086b7b40?, {0x8abbd20?, 0xc00be9c0f0?}, 0xc009a53278?})
	pkg/util/ctxgroup/ctxgroup.go:151 +0xbb
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0x8abbd58?, 0xc001d95320?}, {0xc002c0f068, 0x4, 0x8abbad0?})
	pkg/util/ctxgroup/ctxgroup.go:224 +0x11d
github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).ingest(0xc00a6178d8, {0x8abbd58, 0xc001d95320}, {0x8b2d350, 0xc00158d130})
	pkg/crosscluster/logical/logical_replication_job.go:368 +0x15aa
github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).ingestWithRetries(0xc00a6178d8, {0x8abbd58, 0xc001d95320}, {0x8b2d350, 0xc00158d130})
	pkg/crosscluster/logical/logical_replication_job.go:932 +0x23a
github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).Resume(0xc00a6178d8, {0x8abbd58, 0xc001d95320}, {0x7634320?, 0xc00158d130})
	pkg/crosscluster/logical/logical_replication_job.go:115 +0x11e
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0xc003225980?, 0x0?, 0xc00973f8e0, {0x8a907a0?, 0xc00a6178d8?}, {0x8abbd58?, 0xc001d95320?}, {0x7634320?, 0xc00158d130?})
	pkg/jobs/registry.go:1647 +0xb4
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0083aa008, {0x8abdf40, 0xc00a917400}, {0x7634320, 0xc00158d130}, {0x8a907a0, 0xc00a6178d8}, 0xc003225980, {0xc005aa4412, 0x7}, ...)
	pkg/jobs/registry.go:1648 +0xb7f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc0083aa008, {0x8abdf40, 0xc00a917400}, {0x8a907a0, 0xc00a6178d8}, 0xc003225980, {0xc005aa4412, 0x7}, {0xc009977e48, 0x17})
	pkg/jobs/adopt.go:418 +0x4f9
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1({0x8abdf40?, 0xc00a917220?})
	pkg/jobs/adopt.go:271 +0x13c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x8abdf40, 0xc00a917220})
	pkg/util/stop/stopper.go:440 +0x8a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 30032
	pkg/util/stop/stopper.go:438 +0x125
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomStream
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestRandomStream347743445
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:182: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    logical_replication_job_test.go:1007: waiting for 100 events
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #159971 crosscluster/logical: TestRandomStream failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-26.1 release-blocker]
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomStream.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58111

crosscluster/logical.TestRandomStream [failed](https://mesolite.cluster.engflow.com/invocations/default/83ecc41e-d6f3-4c1a-a644-346f2f56b929?testReportRun=7&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY3Jvc3NjbHVzdGVyL2xvZ2ljYWw6bG9naWNhbF90ZXN0) on master @ [9c3c6369cc4ca7c1d52555bcf5ecfb17417fd57a](https://github.com/cockroachdb/cockroach/commits/9c3c6369cc4ca7c1d52555bcf5ecfb17417fd57a):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait
	  | 	pkg/util/ctxgroup/ctxgroup.go:151
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).close
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:461
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).ConsumerClosed
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:437
	  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Close
	  | 	pkg/sql/execinfra/processorsbase.go:779
	  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup
	  | 	pkg/sql/flowinfra/flow.go:678
	  | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup
	  | 	pkg/sql/rowflow/row_based_flow.go:490
	  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run.func3
	  | 	pkg/sql/distsql_running.go:999
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
	  | 	pkg/util/log/logcrash/crash_reporting.go:434
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Next
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:397
	  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
	  | 	pkg/sql/execinfra/base.go:193
	  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
	  | 	pkg/sql/execinfra/processorsbase.go:766
	  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
	  | 	pkg/sql/flowinfra/flow.go:574
	  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
	  | 	pkg/sql/distsql_running.go:1064
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).ingest.func7
	  | 	pkg/crosscluster/logical/logical_replication_job.go:332
	  | [...repeated from below...]
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | runtime.panicmem
	  | 	GOROOT/src/runtime/panic.go:262
	  | runtime.sigpanic
	  | 	GOROOT/src/runtime/signal_unix.go:925
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.getColumnSchema
	  | 	pkg/crosscluster/logical/replication_statements.go:42
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.newEventDecoder.func1
	  | 	pkg/crosscluster/logical/event_decoder.go:82
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).DescsTxn.func1
	  | 	pkg/sql/internal.go:1955
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
	  | 	pkg/sql/internal.go:2052
	  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
	  | 	pkg/kv/txn.go:1139
	  | github.com/cockroachdb/cockroach/pkg/kv.runTxn
	  | 	pkg/kv/db.go:1114
	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
	  | 	pkg/kv/db.go:1077
	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn
	  | 	pkg/kv/db.go:1052
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
	  | 	pkg/sql/internal.go:2039
	  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).DescsTxn
	  | 	pkg/sql/internal.go:1953
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.newEventDecoder
	  | 	pkg/crosscluster/logical/event_decoder.go:75
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.newCrudSqlWriter
	  | 	pkg/crosscluster/logical/sql_crud_writer.go:51
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).setupBatchHandlers
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:729
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).flushBuffer
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:785
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).handleStreamBuffer
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:646
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).handleEvent
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:529
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).consumeEvents
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:502
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Start.func2.1
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:359
	  | github.com/cockroachdb/cockroach/pkg/util/pprofutil.Do
	  | 	pkg/util/pprofutil/labels.go:38
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Start.func2
	  | 	pkg/crosscluster/logical/logical_replication_writer_processor.go:358
	  | github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationWriterProcessor).Start.Group.GoCtx.func10
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_amd64.s:1693
	Wraps: (3) runtime error: invalid memory address or nil pointer dereference
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) runtime.errorString
	panic: found 1 short-living non-stopped monitors in internal-planner.resume-job-1135646089090662401
	internal-planner.resume-job-1135646089090662401 0 B
	    flow-87cfbc66 0 B
	

```
Stack: 

```
goroutine 29873 [running]:
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc010c2bce0, {0x8ad6460, 0xc003a5cd20}, 0x1)
	pkg/util/mon/bytes_usage.go:781 +0x653
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	pkg/util/mon/bytes_usage.go:722
github.com/cockroachdb/cockroach/pkg/sql.newInternalPlanner.func1()
	pkg/sql/planner.go:521 +0x5f
panic({0x7221ee0?, 0xc005c74d60?})
	GOROOT/src/runtime/panic.go:783 +0x132
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc00a0d6c80?, {0x8ad4240?, 0xc00b4a3c70?}, 0xc00ba1df68?})
	pkg/util/ctxgroup/ctxgroup.go:151 +0xbb
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0x8ad4278?, 0xc012787d40?}, {0xc00a2c9068, 0x4, 0x8ad3ff0?})
	pkg/util/ctxgroup/ctxgroup.go:224 +0x11d
github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).ingest(0xc0119d7dd0, {0x8ad4278, 0xc012787d40}, {0x8b45a10, 0xc004527f68})
	pkg/crosscluster/logical/logical_replication_job.go:368 +0x15aa
github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).ingestWithRetries(0xc0119d7dd0, {0x8ad4278, 0xc012787d40}, {0x8b45a10, 0xc004527f68})
	pkg/crosscluster/logical/logical_replication_job.go:932 +0x23a
github.com/cockroachdb/cockroach/pkg/crosscluster/logical.(*logicalReplicationResumer).Resume(0xc0119d7dd0, {0x8ad4278, 0xc012787d40}, {0x7648ec0?, 0xc004527f68})
	pkg/crosscluster/logical/logical_replication_job.go:115 +0x11e
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0xc007162600?, 0x0?, 0xc00d7ed8e0, {0x8aa8c80?, 0xc0119d7dd0?}, {0x8ad4278?, 0xc012787d40?}, {0x7648ec0?, 0xc004527f68?})
	pkg/jobs/registry.go:1647 +0xb4
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc00c255008, {0x8ad6460, 0xc003a5cdc0}, {0x7648ec0, 0xc004527f68}, {0x8aa8c80, 0xc0119d7dd0}, 0xc007162600, {0xc01131bc02, 0x7}, ...)
	pkg/jobs/registry.go:1648 +0xb7f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc00c255008, {0x8ad6460, 0xc003a5cdc0}, {0x8aa8c80, 0xc0119d7dd0}, 0xc007162600, {0xc01131bc02, 0x7}, {0xc00d0da528, 0x17})
	pkg/jobs/adopt.go:418 +0x4f9
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1({0x8ad6460?, 0xc003a5cbe0?})
	pkg/jobs/adopt.go:271 +0x13c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x8ad6460, 0xc003a5cbe0})
	pkg/util/stop/stopper.go:440 +0x8a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 29878
	pkg/util/stop/stopper.go:438 +0x125
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomStream
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestRandomStream1316378213
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:182: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    logical_replication_job_test.go:1007: waiting for 100 events
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #159971 crosscluster/logical: TestRandomStream failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-26.1 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomStream.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -918,7 +918,7 @@ func getDescriptorsFromStoreForInterval(
 //  2. Figure out a sane policy on when these descriptors should be purged.
 //     They are currently purged in PurgeOldVersions.
 func (m *Manager) readOlderVersionForTimestamp(
-	ctx context.Context, id descpb.ID, timestamp hlc.Timestamp,
+	ctx context.Context, id descpb.ID, timestamp ReadTimestamp,
 ) ([]historicalDescriptor, error) {
 	// Retrieve the endTimestamp for our query, which will be the first
 	// modification timestamp above our query timestamp.
@@ -949,7 +949,7 @@ func (m *Manager) readOlderVersionForTimestamp(
 		// that represents the start of the validity interval for the known version
 		// which immediately follows the timestamps we're searching for.
 		i := sort.Search(len(t.mu.active.data), func(i int) bool {
-			return timestamp.Less(t.mu.active.data[i].GetModificationTime())
+			return timestamp.GetTimestamp().Less(t.mu.active.data[i].GetModificationTime())
 		})
 
 		// If the timestamp we're searching for is somehow after the last descriptor
@@ -964,7 +964,7 @@ func (m *Manager) readOlderVersionForTimestamp(
 			// If we found a descriptor that isn't the first descriptor, go and check
 			// whether the descriptor for which we're searching actually exists. This
 			// will deal with cases where a concurrent fetch filled it in for us.
-			i > 0 && timestamp.Less(t.mu.active.data[i-1].getExpiration(ctx)) {
+			i > 0 && timestamp.GetTimestamp().Less(t.mu.active.data[i-1].getExpiration(ctx)) {
 			return hlc.Timestamp{}, false, true
 		}
 		return t.mu.active.data[i].GetModificationTime(), false, false
@@ -976,7 +976,7 @@ func (m *Manager) readOlderVersionForTimestamp(
 	// Retrieve descriptors in range [timestamp, endTimestamp) in decreasing
 	// modification time order.
 	descs, err := getDescriptorsFromStoreForInterval(
-		ctx, m.storage.db.KV(), m.Codec(), id, timestamp, endTimestamp, isOffline,
+		ctx, m.storage.db.KV(), m.Codec(), id, timestamp.GetTimestamp(), endTimestamp, isOffline,
 	)
 	if err != nil {
 		return nil, err
@@ -997,9 +997,18 @@ func (m *Manager) readOlderVersionForTimestamp(
 	// ExportRequest, we'll invoke another call to retrieve the descriptor with
 	// modification time prior to the timestamp. For offline descriptors, we will
 	// try reading again if nothing was read above as a last resort.
-	if timestamp.Less(earliestModificationTime) || (isOffline && len(descs) == 0) {
+	if timestamp.GetTimestamp().Less(earliestModificationTime) || (isOffline && len(descs) == 0) {
 		desc, err := m.storage.getForExpiration(ctx, earliestModificationTime, id)
 		if err != nil {
+			// In locked reading timestamp, we will attempt to read with the lease timestamp. If the descriptor
+			// is found to be missing at that time, then it may be readable at our base timestamp.
+			// When we scanned [timestamp, endTimestamp), we implicitly scanned the base timestamp as well,
+			// so check if anything from that timestamp fits.
+			if len(descs) > 0 &&
+				errors.Is(err, catalog.ErrDescriptorNotFound) && timestamp.GetTimestamp() != timestamp.GetBaseTimestamp() &&
+				earliestModificationTime.LessEq(timestamp.GetBaseTimestamp()) {
+				return descs, nil
+			}
 			return nil, err
 		}
 		descs = append(descs, historicalDescriptor{
@@ -2229,7 +2238,7 @@ func (m *Manager) Acquire(
 
 		case errors.Is(err, errReadOlderVersion):
 			// Read old versions from the store. This can block while reading.
-			versions, errRead := m.readOlderVersionForTimestamp(ctx, id, timestamp.GetTimestamp())
+			versions, errRead := m.readOlderVersionForTimestamp(ctx, id, timestamp)
 			if errRead != nil {
 				return nil, errRead
 			}

--- a/pkg/sql/catalog/lease/lease_internal_test.go
+++ b/pkg/sql/catalog/lease/lease_internal_test.go
@@ -1260,7 +1260,7 @@ func TestReadOlderVersionForTimestamp(t *testing.T) {
 			resetDescriptorState(manager, tableID, tc)
 
 			// Retrieve historicalDescriptors modification times.
-			retrieved, err := manager.readOlderVersionForTimestamp(ctx, tableID, tc.ts)
+			retrieved, err := manager.readOlderVersionForTimestamp(ctx, tableID, TimestampToReadTimestamp(tc.ts))
 			require.NoError(t, err)
 
 			// Validate retrieved descriptors match expected versions.

